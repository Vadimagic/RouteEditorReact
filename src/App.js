import { Component } from "react";
import Navbar from "./components/Navbar/Navbar";
import Navigation from "./components/Navigation/Navigation";
import arrayMove from 'array-move';
import './style.css';

class App extends Component{
	state = {
		placemarks: [
			{id: '11893129819110126', title: "Marina", position: [50.751574, 137.073856]},
			{id: '8884471543416042', title: "Vadim", position: [55.751574, 37.573856]},
		],
		centerNow: [55.75, 37.57]
	}

	createMarker = e => {
		e.preventDefault();

		const formData  = Object.fromEntries(new FormData(e.target).entries());
		if (!formData.title.trim().length) {
			return;
		}
		
		let marker = {
			id:(new Date()).getTime()+Math.random(),
			title: formData.title.trim(),
			position: this.state.centerNow,
			index:this.state.placemarks.length
		};

		this.setState({
			placemarks: this.state.placemarks.concat(marker)
		});

		e.target.reset()
	}

	removeMarker = index => {
		const placemarksCopy = this.state.placemarks.filter((_, i) => i !== index);

		this.setState({
			placemarks: placemarksCopy
		});
	} 

	onSortEnd=({oldIndex,newIndex})=>{
		this.setState({
				placemarks:arrayMove(this.state.placemarks, oldIndex, newIndex).map((mark,i)=>{
					mark.index=i;
					return mark;
			})
		});
	};
	
	render() {
		return (
			<div className="container">
				<Navbar 
					placemarks={this.state.placemarks}
					createMarker={this.createMarker}
					removeMarker={this.removeMarker}
					onSortEnd={this.onSortEnd}
				/>
				<Navigation 
					geometry={this.state.placemarks}
				/>
			</div>
		)
	}
}

export default App;
