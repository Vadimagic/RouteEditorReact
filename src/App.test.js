import React from 'react';
import App from './App';
import {shallow} from 'enzyme';

describe('Testing <App/>', () => {
	
	const app = shallow(<App/>);

	describe('Testing snapshot & state', () => {
		it('App have render correctly', () => {
			expect(app).toMatchSnapshot();
		});
	
		it('App state "placemarks" is array', () => {
			expect(app.state().placemarks).toBeArray();
		});
	
		it('App state "center" is array', () => {
			expect(app.state().center).toBeArray();
		});
	});

	describe('Handlers tests', () => {
		it('Testing onBoundsChange', () => {
			const position = [53.11, 37.22];
			app.instance().onBoundsChange({get: () => {return {getCenter: () => position}}});
			expect(app.state().center).toBe(position)
		});

		it('Testing changePlacemarksPosition', () => {
			app.state().placemarks = [
				{id: '11893129819110126', title: "Marina", position: [50.751574, 137.073856]},
				{id: '8884471543416042', title: "Vadim", position: [55.751574, 37.573856]},
			];
			const index = 1, 
					newPosition = [55.75, 37.57];
			app.instance().changePlacemarksPosition(index, newPosition);
			expect(app.state().placemarks[index].position).toBe(newPosition)
		});

		it('Testing onSortEnd', () => {
			app.state().placemarks = [
				{id: '11893129819110126', title: "Marina", position: [50.751574, 137.073856]},
				{id: '8884471543416042', title: "Vadim", position: [55.751574, 37.573856]},
			];
			const oldIndex = 0,
					oldMark = app.state().placemarks[oldIndex],
					newIndex = 1, 
					newMark = app.state().placemarks[newIndex];
			app.instance().onSortEnd({oldIndex, newIndex});
			expect(app.state().placemarks[oldIndex]).toBe(newMark);
			expect(app.state().placemarks[newIndex]).toBe(oldMark);
		});

		it('Testing removeMarker', () => {
			app.state().placemarks = [
				{id: '11893129819110126', title: "Marina", position: [50.751574, 137.073856]},
				{id: '8884471543416042', title: "Vadim", position: [55.751574, 37.573856]},
			];
			const index = 1,
					elem = [app.state().placemarks[index]];
			app.instance().removeMarker(index);
			expect(app.state().placemarks).toEqual(expect.not.arrayContaining(elem));
		});
	});
});