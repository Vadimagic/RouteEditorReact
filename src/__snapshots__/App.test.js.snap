// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <App/> Testing snapshot & state App have render correctly 1`] = `ShallowWrapper {}`;

exports[`Testing <Label/> App have render correctly 1`] = `ShallowWrapper {}`;

exports[`Testing <Label/> Testing snapshot & state App have render correctly 1`] = `ShallowWrapper {}`;
